sdk/clientcore/core/src/main/java/io/clientcore/core/http/models/SdkRequestContext.java

```diff
@@ -0,0 +1,135 @@
+    }
+
+    /**
+     * Creates a new instance of {@link SdkRequestContext} with the given options.
+     * @param options The {@link RequestOptions} to be used.
+     * @return A new instance of {@link SdkRequestContext} or, when optimization is possible, the existing instance of {@link SdkRequestContext}
+     * provided in the {@code options} parameter.
+     */
+    public static SdkRequestContext create(RequestOptions options) {
+        if (options instanceof SdkRequestContext) {
+            return (SdkRequestContext) options;
+        }
+
+        return new SdkRequestContext(options, null);
+    }
+
+    @Override
+    public InstrumentationContext getInstrumentationContext() {
+        if (childInstrumentationContext != null && childInstrumentationContext.isValid()) {
+            return childInstrumentationContext;
+        }
+
+        return super.getInstrumentationContext();
+    }
+
+    /**
+     * Get the {@link ProgressReporter} from the request context.
+     *
+     * @return The {@link ProgressReporter} if present, otherwise null.
+     */
+    public ProgressReporter getProgressReporter() {
+        Object progressReporter = super.getData(PROGRESS_REPORTER_KEY);
+        if (progressReporter instanceof ProgressReporter) {
+            return (ProgressReporter) progressReporter;
+        }
+
+        if (progressReporter != null) {
+            LOGGER.atVerbose()
+                .addKeyValue("actualType", progressReporter.getClass().getCanonicalName())
+                .addKeyValue("contextKey", PROGRESS_REPORTER_KEY)
+                .log("Unexpected object type. Ignoring it.");
+        }
+
+        return null;
+    }
+
+    @Override
+    public RequestOptions putData(String key, Object value) {
+        throwOnImmutable("Cannot put data");
+        return this;
+    }
+
+    @Override
+    public RequestOptions setLogger(ClientLogger logger) {
+        throwOnImmutable("Cannot set logger");
+        return this;
+    }
+
+    @Override
+    public RequestOptions setInstrumentationContext(InstrumentationContext instrumentationContext) {
+        throwOnImmutable("Cannot set instrumentation context");
+        return this;
+    }
+
+    @Override
+    public RequestOptions addRequestCallback(Consumer<HttpRequest> requestCallback) {
+        throwOnImmutable("Cannot add request callback");
+        return this;
+    }
+
+    @Override
+    public RequestOptions addQueryParam(String name, String value) {
+        throwOnImmutable("Cannot add query param");
+        return this;
+    }
+
+    @Override
+    public RequestOptions addQueryParam(String parameterName, String value, boolean encoded) {
+        throwOnImmutable("Cannot add query param");
+        return this;
+    }
+
+    private void throwOnImmutable(String message) {
+        throw LOGGER.logThrowableAsError(
+            new IllegalStateException(message + ". This instance of RequestOptions is immutable."));
```

Comment:
Should we use `SdkRequestContext` here?

Also super minor nit: Could we change this method to create the `IllegalStateException` and have all callers to this `throw` in the method itself? This removes the need to have `return` and helps the stack trace by removing one level by removing this method call.

Reply:
moved throw to the caller.
Also removed `RequestOptions` - we use this in `RequestOptions` and `SdkRequestContext` so it's probably best to remove the specific type from the message.